lib/compiler/ast-to-ops.ts

```diff
@@ -649,6 +649,17 @@ function compileTableToOps(table : Ast.Expression,
             compiled.handle_thingtalk, 
             table
         );
+    } else if (table instanceof Ast.JoinExpression) {
+        const lhsop = compileTableToOps(table.lhs, restrictHintsForJoin(hints, table.lhs.schema!));
+        const rhsop = compileTableToOps(table.rhs, restrictHintsForJoin(hints, table.rhs.schema!));
+        const condition = table.condition ? compileBooleanExpressionToOp(table.condition) : null;
+        let device : Ast.DeviceSelector|null = null;
+        let handle_thingtalk = false;
+        if (lhsop.device && rhsop.device) {
+            device = sameDevice(lhsop.device, rhsop.device) ? lhsop.device : null;
+            handle_thingtalk = sameDevice(lhsop.device, rhsop.device) ? lhsop.handle_thingtalk && rhsop.handle_thingtalk : false;
+        }
+        return new TableOp.Join(lhsop, rhsop, condition, device, handle_thingtalk, table);
```

Comment:
Why do we need a new TableOp? NestedLoopJoin or CrossJoin should have the right semantics, because they were written at the time we had joins in ThingTalk 1.0

In fact, as a far as the compiler is concerned, a join is a CrossJoin op wrapped with a FilterOp, isn't it?

Reply:
The semantic are different. Cross join will merge the scope (conflict fields on the lhs will be dropped), while `join` will keep both with prefix.

Reply:
I think we need to rename this stuff, because it's confusing. Also, CrossJoin for => doesn't really make sense, we should make => always use the nested-loop algorithm.

Reply:
Sure, separate PR is fine. But we really need to fix this or we'll forget what's going on and nobody else will be able to touch this code.